Add ArrayTextureImageBinding support to ComputerShader

shi-weili opened the pull request:
Code is copied from `ShaderGL3.kt` and tested.

I'm not sure if other ImageBinding types also work with compute shader. If so, can also add them.

hamoid commented:
Nice :) I was thinking that it might be great to have a minimal demo using this feature at https://github.com/openrndr/openrndr/tree/master/openrndr-demos/src/main/kotlin but you would need some images for that, right?

shi-weili commented:
@hamoid You don't necessarily need image files for a minimal demo 😄 
I've added `DemoComputeShaderArrayTexture01` to the demo folder.
I also added support for other `ImageBinding` types (all copied from `ShaderGL3`). But I'm not familiar with many of these types, so I don't have good ways to test.

hamoid commented:
Hi :) @edwinRNDR was asking me if I could take a look at this PR and make sure it works with the most recent version of OPENRNDR. What do you think about these changes?

To the demo I added some comments to make what's happening more obvious

```kotlin
import org.intellij.lang.annotations.Language
import org.openrndr.application
import org.openrndr.draw.*

/**
 * This demo creates an array of textures (layers)
 * and draws vertical lines on each layer.
 * The horizontal shift of the lines equals the layer ID.
 *
 * The layers are populated only once using a compute shader
 * when the program starts.
 *
 * Later, inside the animation loop,
 * it draws one of the layers on each frame.
 */
fun main() {
    application {
        val textureSize = 200
        val numLayers = 60

        configure {
            width = textureSize
            height = textureSize
        }

        program {
            val arrayTex = arrayTexture(
                textureSize,
                textureSize,
                numLayers,
                ColorFormat.RGBa,
                ColorType.FLOAT32
            )

            @Language("GLSL")
            val cs = ComputeShader.fromCode("""
                #version 450
                layout(local_size_x = 1) in;
                layout(rgba32f) uniform writeonly image2DArray writeTex;
                uniform int numLayers;
                void main() {
                    for (int i = 0; i < numLayers; i++) {
                        imageStore(
                            writeTex,
                            // Note the ivec3, where z = texture ID = layer
                            ivec3(gl_GlobalInvocationID.x, gl_GlobalInvocationID.y, i),
                            vec4((gl_GlobalInvocationID.x + i) % numLayers == 0 ? 1.0 : 0.0,
                            0.0, 0.0, 1.0));
                    }
                }
            """.trimMargin(), "arrayTextureComp")

            // populate all layers using a compute shader
            cs.uniform("numLayers", numLayers)
            cs.image(
                "writeTex",
                0,
                arrayTex.imageBinding(0, ImageAccess.WRITE))
            cs.execute(textureSize, textureSize)

            extend {
                // draw one of the layers
                val currentLayer = frameCount % numLayers
                drawer.image(arrayTex, currentLayer)
            }
        }
    }
}
```

hamoid commented:
In the `image` function I reduced a bit of duplication:

```kotlin
    override fun image(name: String, image: Int, imageBinding: ImageBinding) {
        require((Driver.instance as DriverGL3).version >= DriverVersionGL.VERSION_4_3)

        fun bindTexture(texture: Int, glFormat: Int) = glBindImageTexture(
            image, texture, imageBinding.level, false, 0,
            imageBinding.access.gl(), glFormat
        )
        fun check(colorFormat: ColorFormat) {
            require(colorFormat.componentCount != 3) {
                "ComputeShader image `$name` has unsupported format " +
                        "($colorFormat), only formats with 1, 2 or 4 " +
                        "components are supported"
            }
        }

        bound {
            when (imageBinding) {
                is BufferTextureImageBinding ->
                    (imageBinding.bufferTexture as BufferTextureGL3).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                is ColorBufferImageBinding ->
                    (imageBinding.colorBuffer as ColorBufferGL3).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                is ArrayTextureImageBinding ->
                    (imageBinding.arrayTexture as ArrayTextureGL3).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                is CubemapImageBinding ->
                    (imageBinding.cubemap as CubemapGL3).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                is ArrayCubemapImageBinding ->
                    (imageBinding.arrayCubemap as ArrayCubemapGL4).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                is VolumeTextureImageBinding ->
                    (imageBinding.volumeTexture as VolumeTextureGL3).apply {
                        check(format)
                        bindTexture(texture, glFormat())
                    }
                else -> error("unsupported binding")
            }
            checkGLErrors()
            val index = uniformIndex(name)
            if (index != -1) {
                glUniform1i(index, image)
                checkGLErrors()
            }
        }
    }
```

I'm not 100% sure about the last if. All the `uniform` methods in the same file have this check, so I thought it might be good to copy it.

I tried with OPENRNDR from git and the demo works. Also my own compute shader programs using `image` as input also work as expected after applying these changes.

hamoid commented:
I tried provoking an error by using `ColorFormat.RGB` in the demo instead of `ColorFormat.RGBa` but the expected error message triggered by `require` is only visible if `-Dorg.openrndr.exceptions=JVM` is used in the VM options, otherwise it shows an `indexOutOfBounds`.

hamoid commented:
I tried applying these changes to master but it doesn't work as expected anymore. 

In the demo it seems like only layer 0 is populated correctly, the other 59 layers are transparent.

hamoid commented:
Thank you @shi-weili for your PR. I will try to get it merged soon as https://github.com/openrndr/openrndr/pull/369 (giving you credit for your work).

Feedback / code / suggestions are welcome :)

shi-weili commented:
Thank you @hamoid for the work! :)
